XWIKI-17664: When adding an annotation the focus should be inside WYSIWYG by default

Sereza7 opened the pull request:
# Jira URL

<!-- Add the link to the corresponding JIRA issue referenced in a commit message. Unless this is a [Misc] commit,
see https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HRule:Don27tcreateunnecessaryissues
-->
https://jira.xwiki.org/browse/XWIKI-17664

# Changes

## Description

<!-- Describe the main changes brought in this PR. -->
* Added a comment to explain a line of code that doesn't work with CKEditor anymore
* Added a CKE config change so that it gets focus when loaded.

## Clarifications

<!-- Provide extra hints to make it easier to understand the PR. Those could be:
* Explanation of choices made in this PR
* Anchor towards extra resources needed to understand the context of this PR (e.g., a forum proposal).
* Links to other issues this issue depends on
-->

* The legacy way was just a textarea for the annotation edition. I left this in for backwards compatibility. However, in parallel I added a CKEditor config change that would focus the field once it's loaded. 
* Comments explain why I went with this solution precisely.

# Screenshots & Video

<!-- If this PR introduces any UI change, it's recommended to highlight it with before/after screenshots 
or even a screen recording for complex interactions. 
-->
On the video below, we can see a basic use case and experimentation with adjacent UIs (namely inline editing and comment edition).

https://github.com/user-attachments/assets/36185532-a0cc-45f5-8818-606862570da1

# Executed Tests

<!-- Especially important for regression fixes. 
Indicate how changes were tested (e.g., what maven commands were run to validate them).
-->
Manual tests, see above. 
I doubt the changes would change anything to tests since it's only a minor CKEditor config update (and the docker tests pretty much don't care about the focus state).
Built changes with `mvn clean install -f xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui` and then successfully passed the docker test suite: `mvn clean install -f xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-test/xwiki-platform-annotation-test-docker`.

# Expected merging strategy

* Prefers squash: Yes <!-- No — Explain why. -->
* Backport on branches:
  * 16.10.X

Sereza7 commented:
Assigning surli to this ticket since he reviewed and merged https://github.com/xwiki/xwiki-platform/pull/3239 which was a very similar solution.

surli commented on xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Script.xml:
```diff
@@ -1677,6 +1677,12 @@ XWiki.Annotation = Class.create({
+            // If the WYSIWYG editor used is CKEditor, we make sure its input gets focused as soon as it's loaded
+            // Note that this CKE config is easier than having to handle custom eventListeners.
+            // As of 17.0.0, this does not interfere with the inline editor, so this solution is okay.
+            if(typeof CKEDITOR !== 'undefined') {
```
@mflorea I need your input there, is that the proper way to detect CKEditor?

Sereza7 replied:
Note that this is the same as the way we used at https://github.com/xwiki/xwiki-platform/commit/3b2b6f39ae799438f316a59717e49c04d763e871 . If this is not a correct way, we would need to fix that place too.

According to tests on my local instance this works.

surli replied:
Hmm actually in the link you provided @mflorea made some comments about this check and about startupFocus. 
Feels like it could be better to put that in an internal library that we would reuse where needed, and so if we need or want to change the way to perform this, we wouldn't need to check everywhere. 
wdyt? does it worth it?

mflorea replied:
As I said on my comment for https://github.com/xwiki/xwiki-platform/commit/3b2b6f39ae799438f316a59717e49c04d763e871 , I think it's wrong to hard-code the CKEditor usage. The solution should be generic, for any WYSIWYG editor:

* either by setting some attribute on the text area that is enhanced by CKEditor, before the WYSIWYG editor is loaded, and having some code in a CKEditor plugin that reacts to that attribute
* or by triggering some generic event that a CKEditor plugin can listen to and react
* or, if the previous two options are not possible, introduce some generic WYSIWYG editor JavaScript API that we implement for now in a CKEditor plugin, but this requires more work

Sereza7 commented:
In the last few commits, I implemented a new solution:
* Before calling the `displayEdit` on the field, the velocity template sets the variable `$startupFocus` to true.
* The CKEditor default template in EditSheet checks this value, and update the textarea that will be replaced by adding it an appropriate `data-xxx` attribute.
* The extension to the CKEditor methods `extend` and `inline` are now sensitive to the presence of this `data-xxx` attribute and change the CKEditor config accordingly.

I added comments in the annotation and comment module templates to explain what is this weird assignation. The comments refer to CKEditor for the sake of clarity because as of now it's the only implementation of an editor using this config. The extra cost for other editors is only setting a boolean that won't be used.

@mflorea Does this implementation look okay to you?

____
Note that I included a fix on the code that was not in the PR initially. We want to keep things consistent, IMO it makes sense to include it in this very PR since it's not a lot of lines.

mflorea commented on xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:
```diff
@@ -218,9 +218,11 @@ define('xwiki-ckeditor', [
 
     var config = {
       filebrowserUploadUrl: uploadDisabled ? '' : getUploadURL(sourceDocument, 'filebrowser'),
+      startupFocus: startupFocus,
```
```suggestion
      startupFocus,
```

mflorea commented on xwiki-platform-core/xwiki-platform-annotation/xwiki-platform-annotation-ui/src/main/resources/AnnotationCode/Macros.xml:
```diff
@@ -356,6 +356,8 @@
 #macro(displayAnnotationProperty $annotatedDocument $annotationObject $annotationProperty $mode)
   #if ($mode == 'edit' || $mode == 'create')
+    ## This startupFocus parameter is used by the CKEditor WYSIWYG editor.
+    #set ($startupFocus = true)
```
I'm wondering if we shouldn't make his a bit more generic / extensible to prepare the ground for other configuration parameters we may need in the future:

```suggestion
    #set ($wysiwygEditorConfig = {
      'startupFocus': true
    })
```

This will make the configuration more explicit, and it will reduce the risk of naming conflicts with other entries from the script context. WDYT?

surli commented on xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-ui/src/main/resources/CKEditor/EditSheet.xml:
```diff
@@ -181,9 +181,11 @@ define('xwiki-ckeditor', [
 
     var config = {
       filebrowserUploadUrl: uploadDisabled ? '' : getUploadURL(sourceDocument, 'filebrowser'),
+      startupFocus,
```
shouldn't it be `startupFocus: startupFocus`? it's a key: value config object no?

mflorea replied:
It's called "shorthand property name" see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#property_definitions (syntax sugar to reduce boilerplate).

surli replied:
ok thanks for the information, I'm not sure I'd have changed it since IMO it's not helping much reading but that's fine :)

mflorea commented on xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/commentfield.vm:
```diff
@@ -72,8 +72,10 @@
 #initRequiredSkinExtensions()
-  ## This startupFocus parameter is used by the CKEditor WYSIWYG editor.
-  #set ($startupFocus = true)
+## This startupFocus parameter is used by the CKEditor WYSIWYG editor.
```
```suggestion
  ## This startupFocus parameter is used by the CKEditor WYSIWYG editor.
```

Wrong indentation.

Sereza7 replied:
Thanks! Fixed in https://github.com/xwiki/xwiki-platform/pull/3694/commits/27ead97f96c7f0baedfe2cd371f711486cbe5ed8 :+1:

mflorea reviewed (approved):
Looks good, with just a small indentation issue.
